fog-updater/src/fog_update.py

```diff
@@ -0,0 +1,252 @@
+* Please double-check that none of the files are _deleted_ from the list.
+* Delete this branch after merging or closing the PR.
+
+---
+
+The source code of this automation bot lives in [badboy/fog-update-bot](https://github.com/badboy/fog-update-bot).
+"""  # noqa
+
+
+def ts():
+    return str(datetime.datetime.now())
+
+
+def eval_extract(code):
+    """
+    Eval `code` and return a map of variables and their values.
+
+    `code` should be valid Python code.
+    Only the builtins `list` and `set` are provided.
+
+    Note: this execute arbitrary Python code.
+    Because of the limited builtins list this should be reasonably safe.
+    Still only use this with known valid code!
+    """
+
+    # Allow `list` and `set`, so `list(set(a+b+c))` works.
+    globs = {"__builtins__": {"list": list, "set": set, "sorted": sorted}}
+    exec(code, globs)
+    globs.pop("__builtins__")
+    return globs
+
+
+def swap_file_list(content, app, files, metrics_or_pings, library=False):
+    """
+    Replace the list of `metrics_files` or `ping_files` in `content` with `files`
+    for the given app or library..
+    Returns the changed content.
+
+    All other content is left untouched.
+    YAML syntax is assumed.
+    File entries are correctly indented.
+    """
+    output = io.StringIO()
+    state = None
+    if library:
+        app = f"- library_name: {app}"
+    else:
+        app = f"- app_name: {app}"
+    indent = 0
+
+    lines = content.split("\n")
+
+    # Remove trailing newlines.
+    while not lines[-1]:
+        lines.pop()
+
+    for line in lines:
+        if state is None and line.strip() == app:
+            state = "app"
+        elif (
+            state == "app"
+            and metrics_or_pings == "metrics"
+            and "metrics_files:" in line
+        ):
+            state = "files"
+        elif state == "app" and metrics_or_pings == "pings" and "ping_files:" in line:
+            state = "files"
+        elif state == "files":
+            if line.strip().startswith("-"):
+                ws, _ = line.split("-")
+                indent = len(ws)
+                continue
+            else:
+                for file in files:
+                    print(" " * indent, file=output, end="")
+                    print(f"- {file}\n", file=output, end="")
+                state = None
+
+        print(line, file=output)
+
+    return output.getvalue()
+
+
+def get_latest_metrics_index():
+    r = requests.get(INDEX_URL)
+    r.raise_for_status()
+    return r.text
+
+
+def _rewrite_repositories_yaml(repo, branch, data):
+    contents = repo.get_contents("repositories.yaml", ref=branch)
+    content = contents.decoded_content.decode("utf-8")
+
+    new_content = content
+    for item in data:
+        name, metrics_or_pings, library, files = item
+        new_content = swap_file_list(
+            new_content, name, files, metrics_or_pings, library
+        )
+
+    if content == new_content:
+        raise Exception(
```

Comment:
The Software Engineer in me objects to using exceptions to mark unexceptional (common/normal) situations.

The pragmatic programmer in me doesn't see that it makes any difference, though.